### PR TITLE
[WIP] ci: [4.3 only] release with OTP 23.3

### PR DESCRIPTION
**Pull request by qzhuyan**
rpm/deb: build packages with two tracks: 23.2 and 23.3
    - Track 1: Add 23.3, bring in fix of https://github.com/erlang/otp/issues/5346
    - Track 2: Keep 23.2 for the test

The `relup` file generated by 23.3 & 23.2 should be identical. 

docker:
    Default to 23.3

### CODE REVIEW

File: .github/workflows/build_packages.yaml [hunk 1]
```diff
@@ -289,11 +292,21 @@ jobs:
...
-        PROFILE: ${{ matrix.profile}}
+        PROFILE: ${{ matrix.profile }}
+        ERL_OTP: erl${{ matrix.otp_version }}
+        ARCH: ${{ matrix.arch }}
       run: |
         if [ -d /tmp/packages/$PROFILE ]; then
           cd /tmp/packages/$PROFILE
             for var in $(ls emqx-* ); do
+              if [[ $ERL_OTP == erl23.2* ]]; then
+                # Keep package with new OTP as default
```
[the remark this answers is not in the document]
Comment (qzhuyan) [reply .github/workflows/build_packages.yaml:303]: That was our original idea, yes. but I think it is better to let the new downloads get the one with new OTP keeps us away from that bug.  I am open to change it back but ping @turtleDeng,  what do you think?

File: build [hunk 2]
```diff
@@ -94,6 +94,14 @@ make_relup() {
     fi
     RELX_BASE_VERSIONS="$(IFS=, ; echo "${releases[*]}")"
     export RELX_BASE_VERSIONS
+    if [[ ${PKG_VSN} == 4.3* ]]; then
+        echo "EMQX 4.3 specific, overwrite OTP app versions"
+        local emqx_rel_file="${releases_dir}/${PKG_VSN}/emqx.rel"
+        if [ ! -f "${emqx_rel_file}" ]; then
```
Comment (qzhuyan) [inline build:100]: maybe we don't need this if guard

File: rebar.config [hunk 3]
```diff
@@ -63,6 +63,7 @@
     , {mongodb, {git,"https://github.com/emqx/mongodb-erlang", {tag, "v3.0.13"}}}
     , {epgsql, {git, "https://github.com/emqx/epgsql.git", {tag, "4.6.0"}}}
     , {grpc, {git, "https://github.com/emqx/grpc-erl", {tag, "0.6.6"}}}
+    , {wolff, {git, "https://github.com/kafka4beam/wolff", {tag, "1.5.6"}}}
```
Comment (zmstone) [inline rebar.config:66]: this should be a different PR ?

File: scripts/emqx_rel_otp_app_overwrite.escript [hunk 4]
```diff
@@ -0,0 +1,57 @@
+#!/usr/bin/env escript
```
Comment (zmstone) [inline scripts/emqx_rel_otp_app_overwrite.escript:1]: this file is old and should be deleted?

**Comment by qzhuyan** [comment]
I messed up with rebase will fix

File: rebar.config (hunk 3 above)
Comment (qzhuyan) [reply rebar.config:66]: This is part of test commit and will be revert later, I just want to compare the relup file

**Comment by qzhuyan** [comment]
I am stuck at 
```
#12 116.8 ./emqx/bin/emqx: line 45: /emqx/_upgrade_base/emqx/erts-11.2.2.8/bin/erl: Permission denied
[3007](https://github.com/qzhuyan/emqx/runs/7469936249?check_suite_focus=true#step:5:3008)
#12 116.8 FATAL: Unable to start Erlang.
[3008](https://github.com/qzhuyan/emqx/runs/7469936249?check_suite_focus=true#step:5:3009)
#12 116.8 Please make sure openssl-1.1.1 (libcrypto) and libncurses are installed.
[3009](https://github.com/qzhuyan/emqx/runs/7469936249?check_suite_focus=true#step:5:3010)
#12 116.8 Also ensure it's running on the correct platform,
[3010](https://github.com/qzhuyan/emqx/runs/7469936249?check_suite_focus=true#step:5:3011)
#12 116.8 this EMQX release is built for 23.3.4.9-3-x86_64-unknown-linux-gnu-64-ubuntu20.04
[3011](https://github.com/qzhuyan/emqx/runs/7469936249?check_suite_focus=true#step:5:3012)
#12 116.9 + '[' '' = 4.3.16-cffc81d3 ']'
[3012](https://github.com/qzhuyan/emqx/runs/7469936249?check_suite_focus=true#step:5:3013)
#12 116.9 + exit 1
[3013](https://github.com/qzhuyan/emqx/runs/7469936249?check_suite_focus=true#step:5:3014)
------
```

**Comment by qzhuyan** [comment]
Now it could pass the relup test.

https://github.com/qzhuyan/emqx/actions/runs/2738171351

Will squash the commits after review

File: scripts/emqx_rel_otp_app_overwrite.escript [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    ok = file:write_file(RelFile, io_lib:format("~p.", NewData)).
+
+rel_file("emqx-edge", Dir, RelVsn)->
+    rel_file("emqx", Dir, RelVsn);
+rel_file(Profile, Dir, RelVsn)->
+    filename:join([Dir, RelVsn, Profile++".rel"]).
+
+
+%% Couldn't find a good way to get this list dynamicly.
+otp_apps() ->
```
Comment (qzhuyan) [inline scripts/emqx_rel_otp_app_overwrite.escript:44]: move to ext eterm and load it from rebar.confg.erl and this escript.

File: scripts/emqx_rel_otp_app_overwrite.escript [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+%% It simply read OTP app version (incl. ERTS) from the rel file of *NEW* Release ($RelVsn) and write back to the ones
+%% in *OLD* versions ($BASE_VERSIONS)
+%%
+%% note, we use NEW to overwrite OLD is because the modified NEW rel file will be overwritten by next 'rebar relup'
+%%
+main([Dir, Profile, RelVsn, BASE_VERSIONS]) ->
+    {ErtsVsn, Overwrites} = get_otp_apps(rel_file(Profile, Dir, RelVsn), RelVsn),
+    lists:foreach(fun(BaseVer) ->
+                          base_rel_overwrites(BaseVer, Profile, Dir, ErtsVsn, Overwrites)
+                  end, string:tokens(BASE_VERSIONS, ",")).
```
Comment (qzhuyan) [inline scripts/emqx_rel_otp_app_overwrite.escript:14]: add assertion, no restart emulator.

File: .github/workflows/build_packages.yaml [hunk 7]
```diff
@@ -36,7 +36,7 @@ jobs:
           if make emqx-ee --dry-run > /dev/null 2>&1; then
             echo "::set-output name=profiles::[\"emqx-ee\"]"
           else
-            echo "::set-output name=profiles::[\"emqx\", \"emqx-edge\"]"
```
Comment (zmstone) [inline .github/workflows/build_packages.yaml:39]: revert ?

File: .github/workflows/build_packages.yaml [hunk 8]
```diff
@@ -58,7 +58,7 @@ jobs:
   windows:
     runs-on: windows-2019
     needs: prepare
-    if: endsWith(github.repository, 'emqx')
+    if: false #endsWith(github.repository, 'emqx')
```
Comment (zmstone) [inline .github/workflows/build_packages.yaml:61]: revert ?

File: .github/workflows/build_packages.yaml [hunk 9]
```diff
@@ -119,6 +119,7 @@ jobs:
         path: source/_packages/${{ matrix.profile }}/.
 
   mac:
+    if: false
```
Comment (zmstone) [inline .github/workflows/build_packages.yaml:122]: revert ?

File: .github/workflows/build_packages.yaml [hunk 10]
```diff
@@ -304,6 +319,7 @@ jobs:
         path: /tmp/packages/${{ matrix.profile }}/.
 
   docker:
+    if: false
```
Comment (zmstone) [inline .github/workflows/build_packages.yaml:322]: revert?

**Comment by qzhuyan** [comment]
Compared two builds(otp 23.2 and otp 23.3) and relup has the same contents
 https://github.com/qzhuyan/emqx/actions/runs/2775174870

File: scripts/rel_otp_apps.eterm [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+, public_key
+, asn1
+, syntax_tools
+, ssl
+, os_mon
+, inets
+, compiler
+, runtime_tools
+, {mnesia, load}
+, xmerl
```
Comment (zmstone) [inline scripts/rel_otp_apps.eterm:15]: just a note: xmerl is now included in edge edition, the reason why it was not included in edge edition is unknown. this change seems to be fine.
